armi/reactor/flags.py

```diff
@@ -13,57 +13,82 @@
+object's name when constructed; if the name contains any valid flag names, those Flags
+will be assigned to the object. However, specific Flags may be specified within
+blueprints, in which case the name is ignored and only the explicitly-requested Flags
+are applied (see :doc:`/user/inputs/blueprints` for more details).
+
+Individual Flags tend to be various nouns and adjectives that describe common objects
+that go into a reactor (e.g. "fuel", "shield", "control", "duct", "plenum", etc.). In
+addition, there are some generic Flags (e.g., "A", "B", "C", etc.) that aid in
+disambiguating between objects that need to be targeted separately but would otherwise
+have the same Flags. Flags are stored as integer bitfields within the parameter system,
+allowing them to be combined arbitrarily on any ARMI object. Since they are stored in
+bitfields, each new flag definition requires widening this bitfield; therefore, the
+number of defined Flags should be kept relatively small, and each flag should provide
+maximum utility.
+
+Within the code, Flags are usually combined into a "type specification (``TypeSpec``)",
+which is either a single combination of Flags, or a list of Flag combinations. More
+information about how ``TypeSpec`` is interpreted can be found in
+:py:meth:`armi.reactor.composites.ArmiObject.hasFlags`.
+
+Flags are intended to describe `what something is`, rather than `what something should
+do`. Historically, Flags have been used to do both, which has led to confusion. The
+binding of specific behavior to certain Flags should ideally be controlled through
+settings with reasonable defaults, rather than being hard-coded. Currently, much of the
+code still uses hard-coded ``TypeSpecs``, and certain Flags are clearly saying `what
+something should do` (e.g., ``Flags.DEPLETABLE``).
+
+.. note::
+    Flags have a rather storied history. Way back when, code that needed to operate on
+    specific objects would do substring searches against object names to decide if they
+    were relevant. This was very prone to error, and led to all sorts of surprising
+    behavior based on the names used in input files. To improve the situation, Flags
+    were developed to better formalize which strings mattered, and to define canonical
+    names for things. Still almost all flag checks were hard-coded, and
+    aside from up-front error checks, many of the original issues persisted. For
+    instance, developing a comprehensive manual of which Flags lead to which behavior
+    was very difficult.
+
+    Migrating the `meaning` of Flags into settings will allow us to better document how
+    those Flags/settings affect ARMI's behavior.
+
+    As mentioned above, plenty of code still hard-codes Flag ``TypeSpecs``, and certain
+    Flags do not follow the `what something is` convention. Future work should improve
+    upon this as possible.
+
+
+Things that Flags are used for include:
```

Comment:
Should we consider adding some discussion around the use of settings to define the behavior? Also, this is good documentation, but it seems like it should be discoverable in the docs as well. Should we link to this discussion?

Reply:
Yeah there should probably be something in the main doc/ tree pointing to this...